### DISCUSSION TITLE
Spontaneous Scopes in Update Tokens Script

### DISCUSSION THREAD

**Discussion by tawaren** [JanssenProject/jans#7452] [Q&A] [answered]
In our use case prototype, we amongst other things, dynamically bind the lifetime of tokens based on the scopes. The scope represents access levels. For the initial test, we had a few fixed levels (1-3), each having a separate scope, which worked without problems. However, in the end we need a more fine granular approach, like levels from 1-100. Instead of having 100 scopes, we consider using a spontaneous scope. This worked well except for setting the lifetime over an UpdateToken script, as the spontaneous scope does not appear in context.getGrant().getScopes() - Is there an alternative way to get this scope or are spontaneous scopes and  UpdateToken scripts incompatible?

**Comment by ossdhaval** [comment]
I would request @yuriyz to comment on this.

**Comment by yuriyz** [comment]
It should appear if globally and on client level it is properly configured.

Global AS config
- `allowSpontaneousScopes` - should be set to true

Client level config should have :
- `allowSpontaneousScopes` - set to true
- correct `spontaneousScopes` should be set, which matches requested scope

https://docs.jans.io/v1.0.21/admin/auth-server/scopes/#spontaneous-scopes

**Reply by tawaren** [reply]
Thanks for the answer: I have set allowSpontaneousScopes & spontaneousScopeLifetime globally.

During dynamic client registration, I execute:
 RegisterRequest request = context.getRegisterRequest();
 request.setAllowSpontaneousScopes(true);
 request.getSpontaneousScopes().add(scope);

After this, It worked in the PostAuthn script. I thus falsely assumed that the config worked. 

But after inspecting the result over the TUI, the scope shows up under "Client Scripts / Spontaneous Scopes"  instead of "Advanced Client Prop. / spontaneous scopes validation regex" and "Advanced Client Prop. / Allow spontaneos scopes"
After Setting those manually it worked. 

Now I only need to figure out why the dynamic client registration does set the wrong thing.

**Reply by yuriyz** [reply]
Show full request and response from `/register`. Log file `jans-auth.log` in TRACE log level is also welcomed .

**Reply by tawaren** [reply]
Looking at the request-response revealed that the following line seems not to work:
 ```
RegisterRequest request = context.getRegisterRequest();
 request.setAllowSpontaneousScopes(true);
```
The response still reads: "allow_spontaneous_scopes": false

The scope setting line seems to work:
`request.getSpontaneousScopes().add(scope);`
The response reads: "spontaneous_scopes": ["^access:\\d+$"]

Files:
[RegisterInteraction.txt](https://github.com/JanssenProject/jans/files/14010689/RegisterInteraction.txt)
[jans-auth.log](https://github.com/JanssenProject/jans/files/14010774/jans-auth.log)

P.s: Thanks for the quick help, I appreciate it and choosing Janssen over Keycloak proved to be the right decission

**Reply by yuriyz** [reply]
In request there is no `allow_spontaneous_scopes` set.
Request should be
```
Request:
{
	"application_type":"web",
	"scope":"openid email profile",
	"redirect_uris":["https:\/\/client1.interreg.test.org\/auth_callback"],
	"response_types":["code"],
        "allow_spontaneous_scopes": true,
        "spontaneous_scopes": ["^access:\d+$"]
}
```

Did I get it correctly that wrong request is constructed by `RegisterRequest`?
```
RegisterRequest request = context.getRegisterRequest();
request.setAllowSpontaneousScopes(true);
```
Ask because I see parameter applied in latest source code https://github.com/JanssenProject/jans/blob/3be73055531ed65c1bfdc282024699559e7180a8/jans-auth-server/client/src/main/java/io/jans/as/client/RegisterRequest.java#L2157

**Reply by tawaren** [reply]
Sorry if it was not clear, but the following code snippet is part of a dynamic client registration script:
```
RegisterRequest request = context.getRegisterRequest();
request.setAllowSpontaneousScopes(true);
```
If I set "allow_spontaneous_scopes": true manually in the Request, it works, but I would prefer it when that is set in the client registration script, such as all dynamically registered clients have it, and that's the part that currently does not work.

**Reply by yuriyz** [reply, marked as answer]
I see, `context.getClient()` returns server side constructed object. You can change any value there including `allowSpontaneousScopes`.

Like this: `context.getClient().getAttributes().setAllowSpontaneousScopes(true);`
